src/ralph/api/auth/oidc.py

```diff
@@ -0,0 +1,148 @@
+"""OpenID Connect authentication tool for the Ralph API."""
+
+import logging
+from functools import lru_cache
+from typing import Optional, Union
+
+import requests
+from fastapi import Depends, HTTPException, status
+from fastapi.security import OpenIdConnect
+from jose import ExpiredSignatureError, JWTError, jwt
+from jose.exceptions import JWTClaimsError
+from pydantic import BaseModel, Extra
+
+from ralph.api.auth.user import AuthenticatedUser
+from ralph.conf import settings
+
+oauth2_scheme = OpenIdConnect(
+    openIdConnectUrl=f"{settings.RUNSERVER_AUTH_OIDC_ISSUER_URI}"
+    "/.well-known/openid-configuration",
+    auto_error=False,
+)
+
+# API auth logger
+logger = logging.getLogger(__name__)
+
+
+class IDToken(BaseModel):
+    """Pydantic model representing the core of an OpenID Connect ID Token.
+
+    ID Tokens are polymorphic and may have many attributes not defined in the
+    spec thus this model accepts all addition fields.
+
+    Attributes:
+        iss (str): Issuer Identifier for the Issuer of the response.
+        sub (str): Subject Identifier.
+        aud (str): Audience(s) that this ID Token is intended for.
+        exp (int): Expiration time on or after which the ID Token MUST NOT be
+                   accepted for processing.
+        iat (int): Time at which the JWT was issued.
+    """
+
+    iss: str
+    sub: str
+    aud: Optional[str]
+    exp: int
+    iat: int
+
+    class Config:  # pylint: disable=missing-class-docstring # noqa: D106
+        extra = Extra.allow
+
+
+@lru_cache()
+def discover_provider(base_url: str) -> dict:
+    """Discover the authentication server (or OpenId Provider) configuration."""
+    try:
+        response = requests.get(
+            f"{base_url}/.well-known/openid-configuration", timeout=5
+        )
+        response.raise_for_status()
+        return response.json()
+    except requests.exceptions.RequestException as exc:
+        logger.error(
+            "Unable to discover the authentication server configuration: %s", exc
+        )
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Could not validate credentials",
+            headers={"WWW-Authenticate": "Bearer"},
+        ) from exc
+
+
+@lru_cache()
+def get_public_keys(jwks_uri: str):
```

Comment:
Can we be more specific and use Pydantic types?

```suggestion
def get_public_keys(jwks_uri: AnyUrl):
```

Reply:
Can we also type/define a model for returned value?

Reply:
For sure for type hinting the return value, but I didn't define models, I don't know if it is worth the effort and if it won't slow down the auth.